File: linked_lists/middle_of_linked_list.rb
# Definition for singly-linked list:
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val = 0, nxt = nil)
#     @val = val
#     @next = nxt
#   end
# end

def middle_node(head)
  return nil if head.nil?

  slow = head
  fast = head

  while fast && fast.next
    slow = slow.next
    fast = fast.next.next
  end

  slow
end

File: linked_lists/middle_of_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None

        slow = head
        fast = head
        result = []

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow